File: src/TMG.Aimsun/assignment/RoadAssignment.cs
﻿/*
    Copyright 2022 Travel Modelling Group, Department of Civil Engineering, University of Toronto

    This file is part of TMGToolbox for Aimsun.

    TMGToolbox for Aimsun is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    TMGToolbox for Aimsun is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TMGToolbox for Aimsun.  If not, see <http://www.gnu.org/licenses/>.
*/

using System;
using XTMF;

namespace TMG.Aimsun.assignment
{
    [ModuleInformation(Description = "Generate a road assignment")]
    public class RoadAssignment : IAimsunTool
    {
        [ModuleInformation(Description ="SubModule to name the various types of Matrices")]
        public class TrafficClass: IModule
        {
            [RunParameter("Traffic Class Name", "", "Traffic Class Name")]
            public string VehicleType;

            [RunParameter("AIVTT Matrix Name ", "", "Name of the AIVTT Matrix Name")]
            public string AIVTT;

            [RunParameter("Cost Matrix", "", "Name of the Cost Matrix")]
            public string ACostName;

            [RunParameter("Toll Matrix", "", "Name of the Toll Matrix")]
            public string AToll;

            public string Name { get; set; }

            public float Progress => 0f;

            public Tuple<byte, byte, byte> ProgressColour => new Tuple<byte, byte, byte>(120, 25, 100);

            public bool RuntimeValidation(ref string error)
            {
                return true;
            }
        }

        public const string ToolName = "assignment/roadAssignment.py";

        [RunParameter("Name of Scenario", "", "The name of the road assignment scenario")]
        public string ScenarioName;

        [RunParameter("Name of Experiment", "", "The name of the road assignment experiment")]
        public string ExperimentName;

        [RunParameter("Name of Traffic Demand", "", "The name of the Traffic Demand you wish to use for the simulation")]
        public string NameOfTrafficDemand;

        [RunParameter("Name of Public Transit Plan", "", "The name of the public transit plan")]
        public string NameOfPublicTransitPlan;

        [SubModelInformation(Description = "traffic classes", Required = true)]
        public TrafficClass[] TrafficClasses;

        public string Name { get; set; }
        
        public float Progress { get; set; }
        
        public Tuple<byte, byte, byte> ProgressColour => new Tuple<byte, byte, byte>(120, 25, 100);

        public bool RuntimeValidation(ref string error)
        {
            return true;
        }

        public bool Execute(ModellerController aimsunController)
        {
            if (aimsunController == null)
            {
                throw new XTMFRuntimeException(this, "AimsunController is not properly setup or initalized.");
            }
            return aimsunController.Run(this, ToolName,
                JsonParameterBuilder.BuildParameters(writer =>
                {
                    writer.WritePropertyName("scenarioName");
                    writer.WriteValue(ScenarioName);
                    writer.WritePropertyName("experimentName");
                    writer.WriteValue(ExperimentName);
                    writer.WritePropertyName("nameOfTrafficDemand");
                    writer.WriteValue(NameOfTrafficDemand);
                    writer.WritePropertyName("nameOfPublicTransitPlan");
                    writer.WriteValue(NameOfPublicTransitPlan);
                    writer.WritePropertyName("matrixName");
                    writer.WriteStartArray();
                    for (int i = 0; i < TrafficClasses.Length; i++)
                    {
                        writer.WriteStartObject();
                        writer.WritePropertyName("ACostName");
                        writer.WriteValue(TrafficClasses[i].ACostName);
                        writer.WritePropertyName("AIVTT");
                        writer.WriteValue(TrafficClasses[i].AIVTT);
                        writer.WritePropertyName("AToll");
                        writer.WriteValue(TrafficClasses[i].AToll);
                        writer.WritePropertyName("VehicleType");
                        writer.WriteValue(TrafficClasses[i].VehicleType);
                        writer.WriteEndObject();
                    }
                    writer.WriteEndArray();
                }));
        }
    }
}

File: src/TMGToolbox/assignment/roadAssignment.py
"""
    Copyright 2022 Travel Modelling Group, Department of Civil Engineering, University of Toronto

    This file is part of TMGToolbox for Aimsun.

    TMGToolbox for Aimsun is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    TMGToolbox for Aimsun is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with TMGToolbox for Aimsun.  If not, see <http://www.gnu.org/licenses/>.
"""

from PyANGBasic import *
from PyANGKernel import *
from PyANGConsole import *
from PyANGDTA import *
from PyMacroKernelPlugin import *
from PyMacroToolPlugin import *
from PyMacroAdjustmentPlugin import *
from PyMacroPTPlugin import *
from PyFrankWolfePlugin import *
import sys
import commonModule as CM

def create_scenario(model, trafficDemand, ptPlan, xtmf_parameters):
    """
    Function to create the static road assignment scenario
    """
    cmd = model.createNewCmd(model.getType("MacroScenario"))
    model.getCommander().addCommand(cmd)
    scenario = cmd.createdObject()
    scenario.setDemand(trafficDemand)
    scenario.setPublicLinePlan(ptPlan)
    scenario.setName(xtmf_parameters["scenarioName"])
    print(f"Create scenario {scenario.getId()}")

    cmd1 = model.createNewCmd(model.getType("GKPathAssignment"))
    model.getCommander().addCommand(cmd1)
    PathAssignment = cmd1.createdObject()

    dataToOutput = scenario.getOutputData()
    dataToOutput.setGenerateSkims(1)

    scenario.setOutputData(dataToOutput)

    return (scenario, PathAssignment)

def create_experiment_for_scenario(model, scenario, PathAssignment, xtmf_parameters):
    """
    Function to create the experiment
    """
    #create a new experiment
    macroExperimentCmd = model.createNewCmd(model.getType("MacroExperiment"))
    macroExperimentCmd.setScenario(scenario)
    macroExperimentCmd.setEngine("FrankWolfe")
    model.getCommander().addCommand(macroExperimentCmd)
    experiment = macroExperimentCmd.createdObject()
    params = experiment.getParameters()
    params.setMaxIterations(100)
    params.setMaxRelativeGap(0.00001)
    params.setFrankWolfeMethod(CFrankWolfeParams.eNormal)
    #attach the experiment to the scenario
    experiment.setParameters(params)
    model.getCatalog().add(experiment)
    experiment.setScenario(scenario)
    experiment.setOutputPathAssignment(PathAssignment)
    
    return experiment

def deleteExperimentMatrices(model, skimMatrixList, matrixList):
    """
    function to delete experiment output skim matrices
    """
    # loop through the generated output skim matrix list
    for matrix in skimMatrixList:
        # if the name matches don't delete otherwise delete the matrix
        if matrix.getName() in matrixList:
            pass
        else:
            cmd = matrix.getDelCmd()
            model.getCommander().addCommand(cmd)

def buildOriginalAimsunMatrixName(model, experiment_id, parameters, skimMatrixList, catalog):
    """
    Function to create the original aimsun matrix names which will then be renamed 
    in the renameMatrix() function
    """
    matrixList = []
    # matrix prefix for road assignment it is Skim
    matrix_name_prefix = "Skim - "
    # iterate over the matrix list and create new named skim matrices
    for item in parameters:
        # create the names
        ACostName = matrix_name_prefix + "Cost: " + item["VehicleType"] + " " + str(experiment_id)
        AIVTTName = matrix_name_prefix + "Distance: " + item["VehicleType"] + " " + str(experiment_id)
        ATollName = matrix_name_prefix + "tolls: " + item["VehicleType"] + " " + str(experiment_id)
        # rename the matrices
        CM.renameMatrix(model, ACostName, item["ACostName"])
        CM.renameMatrix(model, AIVTTName, item["AIVTT"])
        CM.renameMatrix(model, ATollName, item["AToll"])

        #append the original names to a list for comparison
        matrixList.append(item["ACostName"])
        matrixList.append(item["AIVTT"])
        matrixList.append(item["AToll"])

    # delete the unnamed skim matrices
    deleteExperimentMatrices(model, skimMatrixList, matrixList)

def run_xtmf(parameters, model, console):
    """
    A general function called in all python modules called by bridge. Responsible
    for extracting data and running appropriate functions.
    """
    # extract the parameters and save to dictionary
    xtmf_parameters = {
        "scenarioName": parameters["scenarioName"],
        "experimentName": parameters["experimentName"],
        "trafficDemandName": parameters["nameOfTrafficDemand"],
        "PublicTransitPlanName": parameters["nameOfPublicTransitPlan"],
        "MatrixNames": parameters["matrixName"]
    }
    _execute(model, console, xtmf_parameters)

def _execute(inputModel, console, xtmf_parameters):
    """
    Main execute function to run the simulation
    """
    model = inputModel
    system = GKSystem.getSystem()
    catalog = model.getCatalog()

    # delete all scenarios and experiments
    CM.deleteAimsunObject(model, "MacroExperiment", xtmf_parameters["experimentName"])
    CM.deleteAimsunObject(model, "MacroScenario", xtmf_parameters["scenarioName"])

    # extract the traffic demand object
    trafficDemandObject = catalog.findByName(xtmf_parameters["trafficDemandName"])
    # extract the public transit object
    publicTransitObject = catalog.findByName(xtmf_parameters["PublicTransitPlanName"])
    # Create the scenario
    scenario, pathAssignment = create_scenario(model, trafficDemandObject, publicTransitObject, xtmf_parameters)
    # generate the experiment
    experiment = create_experiment_for_scenario(model, scenario, pathAssignment, xtmf_parameters)
    # Execute the experiment for road assignment
    print("Run road assignment experiment")
    system.executeAction("execute", experiment, [], "static assignment")
    experiment.getStatsManager().createTrafficState()
    # extract the skim matrices 
    skimMatrixList = experiment.getOutputData().getSkimMatrices()
    # get id of experiment
    experiment_id = experiment.getId()
    # build the original aimsun matrix 
    buildOriginalAimsunMatrixName(model, experiment_id, xtmf_parameters["MatrixNames"], skimMatrixList, catalog)
    print ('experiment ran successfully')
    
def runFromConsole(inputArgs):
    """
    This function takes commands from the terminal, creates a console and model to pass
    to the _execute function
    """
    # extract the data from the command line arguments
    inputNetwork = inputArgs[1]
    outputNetworkFile = inputArgs[2]
    # extract the parameters and save to dictionary
    xtmf_parameters = {
        "autoDemand": inputArgs[3],
        # placeholder default values
        "start": float(inputArgs[4]),
        "duration": float(inputArgs[5]),
        "transitDemand": inputArgs[6],
    }
    # Start a console
    console = ANGConsole()
    # generate a model of the input network
    model = None
    # Load a network
    if console.open(inputNetwork):
        model = console.getModel()
        print("Open network")
    else:
        console.getLog().addError("Cannot load the network")
        raise Exception("Cannot load network")

    # call the _execute() function
    _execute(model, console, xtmf_parameters) 

if __name__ == "__main__":
    # function to parse the command line arguments and run network script
    runFromConsole(sys.argv)
